File: BEmu/MarketDataRequest/MarketMessageSubscriptionData.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="MarketDataRequest/MarketMessageSubscriptionData.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MarketMessageSubscriptionData : Message
    {
        private readonly Dictionary<string, Element> _fields;
        private readonly string _security;

        internal MarketMessageSubscriptionData(Subscription sub, Dictionary<string, object> fields) : base(new Name("MarketDataEvents"), sub.CorrelationID, null)
        {
            this._fields = new Dictionary<string, Element>();
            foreach (var item in fields)
            {
                Element elm = null;
                if (item.Value is double)
                    elm = new MarketElementDouble(item.Key, (double)item.Value);

                else if (item.Value is Datetime)
                {
                    Datetime temp = (Datetime)item.Value;

                    bool isDate = temp.HasParts(Datetime.DATE);
                    bool isTime = temp.HasParts(Datetime.TIME);
                    bool isDatetime = isDate && isTime;

                    if (isDatetime)
                        elm = new MarketElementDatetime(item.Key, temp.ToSystemDateTime());
                    else if(isDate)
                        elm = new MarketElementDate(item.Key, temp.ToSystemDateTime());
                    else if(isTime)
                        elm = new MarketElementTime(item.Key, temp.ToSystemDateTime());
                }

                else if (item.Value is int)
                    elm = new MarketElementInt(item.Key, (int)item.Value);
                else if (item.Value is string)
                    elm = new MarketElementString(item.Key, (string)item.Value);
                else if (item.Value is bool)
                    elm = new MarketElementBool(item.Key, (bool)item.Value);

                if (elm != null)
                    this._fields.Add(item.Key, elm);
            }
            this._security = sub.Security;
        }

        public override string TopicName { get { return this._security; } }
        public override IEnumerable<Element> Elements { get { return this._fields.Values; } }
        public override int NumElements { get { return this._fields.Count; } }

        public override object this[string name, int index]
        {
            get
            {
                if (index == 0)
                    return this[name].GetValue();
                else
                    return base[name, index];
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.AppendFormat("{0} = {{{1}", this.MessageType.ToString(), Environment.NewLine);
            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(1));
            }
            result.AppendLine("}");

            return result.ToString();
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/MarketMessageSubscriptionData.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/MarketMessageSubscriptionData.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.Map;
import java.util.HashMap;
import java.util.Map.Entry;

import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Subscription;
import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Message;

public class MarketMessageSubscriptionData extends Message
{
    private final Map<String, Element> _fields;
    private final String _security;

    MarketMessageSubscriptionData(Subscription sub, Map<String, Object> fields) throws Exception
    {
    	super(new Name("MarketDataEvents"), sub.correlationID(), null);
    	
        this._fields = new HashMap<String, Element>();
        
        for(Entry<String, Object> item : fields.entrySet())
        {
            Element elm = null;
            
            if (item.getValue().getClass() == double.class || item.getValue().getClass() == Double.class)
                elm = new MarketElementDouble(item.getKey(), (Double)item.getValue());

            else if (item.getValue().getClass() == Datetime.class)
            {
                Datetime temp = (Datetime)item.getValue();
                
                boolean isDate = temp.hasParts(Datetime.DATE);
                boolean isTime = temp.hasParts(Datetime.TIME);
                boolean isBoth = isDate && isTime;
                
                if(isBoth)
                	elm = new MarketElementDatetime(item.getKey(), temp);
                else if(isDate)
                	elm = new MarketElementDate(item.getKey(), temp);
                else if(isTime)
                	elm = new MarketElementTime(item.getKey(), temp);
            }

            else if (item.getValue().getClass() == int.class ||item.getValue().getClass() == Integer.class)
                elm = new MarketElementInt(item.getKey(), (Integer)item.getValue());
            
            else if (item.getValue().getClass() == String.class)
                elm = new MarketElementString(item.getKey(), (String)item.getValue());
            
            else if (item.getValue().getClass() == boolean.class || item.getValue().getClass() == Boolean.class)
                elm = new MarketElementBool(item.getKey(), (Boolean)item.getValue());

            String key = item.getKey();
            
            if (elm != null)
                this._fields.put(key, elm);
        }
        this._security = sub.security();
    }

    public String topicName()
    {
    	return this._security;
    }
    
    public int numElements()
    {
    	return this._fields.size();
    }
	
	public boolean hasElement(String name, boolean excludeNullElements) throws Exception
	{
		for(Entry<String, Element> item : this._fields.entrySet())
		{
			if(item.getValue().name().toString().equals(name))
				return true;
		}
		return false;
	}
	
	public Element getElement(String name) throws Exception
	{
		for(Entry<String, Element> item : this._fields.entrySet())
		{
			if(item.getValue().name().toString().equals(name))
				return item.getValue();
		}
		return super.getElement(name);
	}
    
	public String toString()
	{
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s = {%s", this.messageType().toString(), System.getProperty("line.separator")));
        
        try
        {
	        for(Entry<String, Element> item : this._fields.entrySet())
	        {
	        	result.append(item.getValue().prettyPrint(1));
	        }
		}
	    catch (Exception e) { }
        result.append(String.format("}%s", System.getProperty("line.separator")));
        
        return result.toString();
	}
    
}
